nb-api/src/main/java/io/nosqlbench/nb/api/nbio/ResolverForNBIOCache.java

```diff
@@ -0,0 +1,321 @@
+        Path path = resolvePath(uri);
+
+        if (path != null) {
+            contents.add(new PathContent(path));
+        }
+        return contents;
+    }
+
+    /**
+     * This method is used to resolve the path of a given URI.
+     * It first checks if the URI has a scheme (http or https) and if it does, it tries to resolve the path from the cache.
+     * If the file is not in the cache, it tries to download it from the remote URL.
+     * If the URI does not have a scheme, it returns null.
+     *
+     * @param uri the URI to resolve the path for
+     * @return the resolved Path object, or null if the URI does not have a scheme or the path could not be resolved
+     */
+    private Path resolvePath(URI uri) {
+        if (uri.getScheme() != null && !uri.getScheme().isEmpty() &&
+            (uri.getScheme().equalsIgnoreCase("http") ||
+                uri.getScheme().equalsIgnoreCase("https"))) {
+            Path cachePath = Path.of(cacheDir + uri.getPath());
+            if (Files.isReadable(cachePath)) {
+                return pathFromLocalCache(cachePath, uri);
+            }
+            else {
+                return pathFromRemoteUrl(uri);
+            }
+        }
+        return null;
+    }
+
+    private boolean downloadFile(URI uri, Path cachePath, URLContent checksum) {
+        int retries = 0;
+        boolean success = false;
+        while (retries < maxRetries) {
+            try {
+                if (this.remoteFileExists(uri)) {
+                    logger.info(() -> "Downloading remote file " + uri + " to cache at " + cachePath);
+                    ReadableByteChannel channel = Channels.newChannel(uri.toURL().openStream());
+                    FileOutputStream outputStream = new FileOutputStream(cachePath.toFile());
+                    outputStream.getChannel().transferFrom(channel, 0, Long.MAX_VALUE);
+                    outputStream.close();
+                    channel.close();
+                    logger.info(() -> "Downloaded remote file to cache at " + cachePath);
+                    if(checksum == null || verifyChecksum(cachePath, checksum)) {
+                        success = true;
+                        break;
+                    }
+                } else {
+                    logger.error(() -> "Error downloading remote file to cache at " + cachePath + ", retrying...");
+                    retries++;
+                }
+            } catch (IOException e) {
+                logger.error(() -> "Error downloading remote file to cache at " + cachePath + ", retrying...");
+                retries++;
+            }
+        }
+        return success;
+    }
+
+    private boolean verifyChecksum(Path cachePath, URLContent checksum) {
+        try {
+            String localChecksumStr = generateSHA256Checksum(cachePath.toString());
+            Path checksumPath = Path.of(cachePath.toString().substring(0, cachePath.toString().lastIndexOf('.')) + ".sha256");
```

Comment:
Some checksum approaches append the `.sha256` to the whole filename without stripping the outermost suffix.

Reply:
Suppose at some point we have `textembedding.hdf5` and `textembedding.ivec` in the nbio cache? We would have a `.sha256` filename collision.

Reply:
yeah that's a really good point. Updated code to simply append '.sha256' to complete filename / url